File: leetcode_alg/solution148.cpp
/**
 * @file solution148.cpp
 * @author your name (you@domain.com)
 * @brief 排序链表
 * @version 0.1
 * @date 2022-09-05
 *
 * @copyright Copyright (c) 2022
 *
 */
#include <iostream>
#include <algorithm>
#include "listnode.h"

using namespace std;

class Solution {
public:
    ListNode *sortList(ListNode *head) {
        if (head == nullptr || head->next == nullptr) {
            return head;
        }
        vector<int> array;
        ListNode* p = head;
        while(p != nullptr){
            array.push_back(p->val);
            p = p->next;
        }
        std::sort(array.begin(), array.end());
//        sort(array, 0, array.size()-1);
        p = head;
        int i = 0;
        while (p != nullptr){
            p->val = array[i++];
            p = p->next;
        }
        return head;
    }

private:
    void sort(vector<int> &array, int left, int right){
        if(left >= right){
            return;
        }
        int l = left;
        int r = right;
        int baseVal = array[left];
        while(l < r){
            // 从右往左找第一个比baseVal小的值
            while(r > l && array[r] >= baseVal){
                r--;
            }
            array[l] = array[r];
            // 从左往右找第一个比baseVal大的值
            while(l < r && array[l] <= baseVal){
                l++;
            }
            array[r] = array[l];
        }
        array[l] = baseVal;
        sort(array, left, l-1);
        sort(array, l+1, right);
    }
};

int main() {
    auto s = new Solution;
    vector<int> array = {4,19,14,5,-3,1,8,5,11,15};
    ListNode *head = buildList(array);
    ListNode *res = s->sortList(head);
    print_list(res);
    free_list(head);
    delete s;
    return 0;
}

File: golang/solution148.go
package main

import (
	"fmt"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func sortList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	start := head

	for {
		if start == nil {
			break
		}
		var next *ListNode = start.Next

		for {
			if next == nil {
				break
			}
			if next.Val < start.Val {
				tmp := next.Val
				next.Val = start.Val
				start.Val = tmp
			}
			next = next.Next
		}
		start = start.Next
	}

	return head
}

func sortList2(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	var slice1 []int
	start := head

	for start != nil {
		slice1 = append(slice1, start.Val)
	}

	return head
}

func sort(array []int, left int, right int) {
	if left >= right {
		return
	}
	var baseVal int = array[right]
	var l int = left
	var r int = right
	for l < r {
		for ; l < r && array[l] < baseVal; l++ {
		}
		array[r] = array[l]
		for ; l < r && array[r] > baseVal; r-- {
		}
		array[l] = array[r]
	}
	array[l] = baseVal
	sort(array, left, l-1)
	sort(array, l+1, right)
}

func main148() {

	var a []int = []int{5, 7, 2, 4, 6}
	sort(a, 0, 4)
	for ix := 0; ix < 5; ix++ {
		fmt.Println(a[ix])
	}

	head := &ListNode{Val: 4, Next: &ListNode{Val: 3, Next: &ListNode{Val: 2, Next: &ListNode{Val: 1, Next: nil}}}}

	sortList(head)

	for {
		if head == nil {
			break
		}
		fmt.Printf("值：%d\n", head.Val)
		head = head.Next
	}
}
